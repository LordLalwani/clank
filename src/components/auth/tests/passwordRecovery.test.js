import React from "react";
import {
  render,
  queryByAttribute,
  cleanup,
  fireEvent,
  waitFor,
} from "@testing-library/react";
import PasswordRecovery from "../passwordRecovery";

import Amplify, { Auth } from "aws-amplify";
import awsconfig from "../../../aws-exports";
import renderer from "react-test-renderer";

Amplify.configure(awsconfig);

const getById = queryByAttribute.bind(null, "id");
afterEach(cleanup);

let DOM;
let emailTextField;
let emailButton;

beforeEach(() => {
  DOM = render(<PasswordRecovery />);
  emailTextField = getById(DOM.container, "email-textfield");
  emailButton = getById(DOM.container, "email-button");
});

describe("render tests", () => {
  it("should render all sub components correctly", () => {
    expect(emailTextField).toBeTruthy();
    expect(emailButton).toBeTruthy();
  });

  it("should render PasswordRecovery component correctly", () => {
    const tree = renderer.create(<PasswordRecovery />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Behavior tests", () => {
  it("should enable email button if email textfield is not empty", () => {
    expect(emailButton).toHaveAttribute("disabled");
    fireEvent.change(emailTextField, { target: { value: "test@test.com" } });
    expect(emailButton).not.toHaveAttribute("disabled");
  });
  it("should show new password & confirmation code textfields and confirm password button upon email button clicked", async () => {
    Auth.forgotPassword = jest.fn().mockImplementation(() => {
      const response = { user: "larryTheLama" };
      return response;
    });
    fireEvent.change(emailTextField, { target: { value: "test@test.com" } });
    expect(emailButton).toBeTruthy();
    fireEvent.click(emailButton);

    await waitFor(() => {
      expect(Auth.forgotPassword).toHaveBeenCalledTimes(1);
    });

    // Was having problems targeting confirmationCodeButton so queried by autoComplete
    // this should be replaced asap >:[
    const getByAutoComplete = queryByAttribute.bind(null, "autoComplete");
    const confirmationCodeTextField = getByAutoComplete(
      DOM.container,
      "confirm-code-textfield"
    );
    expect(confirmationCodeTextField).toBeTruthy();

    expect(DOM.queryByText("Confirm New Password")).toBeTruthy();

    const newPasswordTextfield = getByAutoComplete(
      DOM.container,
      "new-password-textfield"
    );
    expect(newPasswordTextfield).toBeTruthy();
  });
});
