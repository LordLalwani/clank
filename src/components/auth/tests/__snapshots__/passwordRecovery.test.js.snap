// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render tests should render signIn component correctly 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "height": "90vh",
    }
  }
>
  <main
    className="MuiContainer-root MuiContainer-maxWidthXs"
  >
    <div
      className="PasswordRecovery-paper-12"
    >
      <div
        className="MuiAvatar-root MuiAvatar-circle PasswordRecovery-avatar-13 MuiAvatar-colorDefault"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
          />
        </svg>
      </div>
      <h1
        className="MuiTypography-root MuiTypography-h5"
      >
        Recover Your Password
      </h1>
      <form
        className="PasswordRecovery-form-14"
        noValidate={true}
      >
        <div
          className="MuiFormControl-root MuiTextField-root PasswordRecovery-textField-15 MuiFormControl-marginNormal MuiFormControl-fullWidth"
          onClick={[Function]}
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
            data-shrink={false}
            htmlFor="email-textfield"
            id="email-textfield-label"
          >
            Email Address
            <span
              aria-hidden={true}
              className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoComplete="email"
              autoFocus={true}
              className="MuiInputBase-input MuiOutlinedInput-input"
              disabled={false}
              id="email-textfield"
              name="email"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              type="text"
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-17 MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled-19"
              >
                <span>
                  Email Address
                   *
                </span>
              </legend>
            </fieldset>
          </div>
          
        </div>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained Button-button-21 Mui-disabled MuiButton-fullWidth Mui-disabled"
          disabled={true}
          id="email-button"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="submit"
        >
          <span
            className="MuiButton-label"
          >
            Email me a recovery code
          </span>
        </button>
        <div
          className="MuiGrid-root MuiGrid-container"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
          >
            <a
              className="MuiTypography-root MuiLink-root MuiLink-underlineHover PasswordRecovery-link-16 MuiTypography-body2 MuiTypography-colorPrimary"
              href="/Sign-in"
              onBlur={[Function]}
              onFocus={[Function]}
            >
              Sign in
            </a>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item"
          >
            <a
              className="MuiTypography-root MuiLink-root MuiLink-underlineHover PasswordRecovery-link-16 MuiTypography-body2 MuiTypography-colorPrimary"
              href="/sign-up"
              onBlur={[Function]}
              onFocus={[Function]}
            >
              Sign Up
            </a>
          </div>
        </div>
      </form>
    </div>
    <div
      className="MuiBox-root MuiBox-root-23"
    >
      <p
        className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
      >
        Copyright © 
        <a
          className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit"
          href="https://material-ui.com/"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          Clank
        </a>
         
        2020
        .
      </p>
    </div>
  </main>
</div>
`;
